### PR TITLE
feat: change API._get_responder() 404 behavior to raise falcon.HTTPNotFound()

### PR DESCRIPTION
**Pull request by michaelgugino**
feat: change API._get_responder() 404 behavior to raise falcon.HTTPNotFound()

Class API() currenlty accepts optional method add_error_handler() to override raised exceptions.
By default, API._get_responder() method returns a tuple including
responder = falcon.responders.path_not_found. Change made to API._get_responder() to raise
falcon.HTTPNotFound() in place of returning falcon.responders.path_not_found
This change allows easy override of 404 behavior in the application.

### CODE REVIEW

**Comment by philiptzou** [comment]
Will raising an exception instead of returning a response object in `_get_responder` affects some middlewares?

**Comment by michaelgugino** [comment]
I'm not sure if middleware will be affected or not.  I'm not an expert in this code base, this pull is more of a suggestion/request for new behavior.  This pull should only affect default responder behavior, as best I can tell.

**Comment by kgriffs** [comment]
@philiptzou This patch will skip execution of `process_resource()` middleware methods, although the other middleware methods will still run. This would be a breaking change (currently, `process_resource()` receives `None` for its `resource` arg when no route was found.

File: falcon/api.py [hunk 1]
```diff
@@ -469,10 +469,6 @@ def _get_responder(self, req):
             method was not found in the method_map for the responder,
             the responder callable element of the returned tuple will be
             `falcon.responder.bad_request`.
```
Comment (kgriffs) [inline falcon/api.py]: Seems like if we are going to do this for 404, we should also do it for 400. What do you think?

File: falcon/responders.py [hunk 2]
```diff
@@ -18,11 +18,6 @@
 from falcon.status_codes import HTTP_405
 
 
-def path_not_found(req, resp, **kwargs):
-    """Simply sets responseto "404 Not Found", no body."""
-    resp.status = HTTP_404
```
Comment (kgriffs) [inline falcon/responders.py]: I would prefer keeping this function and modifying it to raise falcon.HTTPNotFound(). That way, we can avoid any unintended side-effects; `API.__call__()` is currently written in way that assumes `API._get_responder()` does not raise an error.

**Comment by michaelgugino** [comment]
@kgriffs I implemented your suggested changes, including similar handling of 400, and as a plus the build is now passing.

**Comment by yohanboniface** [comment]
@michaelgugino I think you should rebase :)

File: falcon/api.py [hunk 3]
```diff
@@ -465,12 +465,7 @@ def _get_responder(self, req):
         Note:
             If a responder was matched to the given URI, but the HTTP
             method was not found in the method_map for the responder,
-            the responder callable element of the returned tuple will be
```
Comment (kgriffs) [inline falcon/api.py]: Would you please revert this change to the docstring? `_get_responder` does not call the responder itself, but just returns it.

File: falcon/responders.py [hunk 4]
```diff
@@ -13,19 +13,19 @@
 # limitations under the License.
 
 from falcon.status_codes import HTTP_204
-from falcon.status_codes import HTTP_400
-from falcon.status_codes import HTTP_404
 from falcon.status_codes import HTTP_405
+from falcon.errors import HTTPNotFound
```
Comment (kgriffs) [inline falcon/responders.py]: Nitpick: Not a big deal, but it can be nice to order imports alphabetically.

Comment (yohanboniface) [inline falcon/responders.py]: https://github.com/timothycrosley/isort is a nice tool to handle this :)

**Comment by kgriffs** [comment]
@michaelgugino I'd like to get this finished up and merged soon. Do you have some time to take a look over the next couple of days?

**Comment by michaelgugino** [comment]
Sure, I can make these changes by the end of the week.  I'm a little busy today and tomorrow, possibly Thursday or Friday I'll send up the changes rebased.

**Comment by kgriffs** [comment]
Awesome, thanks!

**Comment by kgriffs** [comment]
Rebased and merged as #696
